[extension/awslogsencodingextension] Support CloudTrail log events

jsvd opened the pull request:
#### Description

This changeset follows the same structure as vpc flow log, s3 access log and subscription filter logs but now dedicated to CloudTrail log events.

#### Link to tracking issues

This is part of the effort described in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/38627:

>  It would be possible to configure the extension with the specific format, such as CloudWatch log group subscription filter, CloudTrail, etc. By having them co-located, we can avoid proliferation of extensions, and better ensure consistency across the formats (e.g. set cloud.* SemConv fields consistently).

#### Testing

The log samples were taken from
https://docs.aws.amazon.com/awscloudtrail/latest/userguide/cloudtrail-log-file-examples.html#cloudtrail-log-file-examples-section.

This unmarshaller is tested against startInstances and createUser events.

#### Documentation

Updated documentation to reflect new CloudTrail Logs unmarshalling.

jsvd committed:
initial commit for unmarshaller of CloudTrail log events

This changeset follows the same structure as vpc flow log, s3 access log
and subscription filter logs but now dedicated to CloudTrail log events.

The log samples were taken from
https://docs.aws.amazon.com/awscloudtrail/latest/userguide/cloudtrail-log-file-examples.html#cloudtrail-log-file-examples-section.

This unmarshaller is tested against startInstances and createUser events
and supports compressed messages.

It takes in a compressed JSON object of `{"Records": []}` where the
array contains individual objects.

The outcome is grouped by account id, and region id.

constanca-m commented on extension/encoding/awslogsencodingextension/config.go:
```diff
@@ -79,5 +83,10 @@ func (cfg *Config) Validate() error {
 		return errors.Join(errs...)
 	}
-	return nil
+	for _, format := range supportedLogFormats {
+		if cfg.Format == format {
+			return nil
+		}
+	}
+	return fmt.Errorf("unsupported format %q, expected one of %q", cfg.Format, supportedLogFormats)
```
This part is an extra, covered in the `switch` above. I will review the rest once the PR gets marked for review :)

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/cloudtraillogs/cloudtraillogs.go:
```diff
@@ -0,0 +1,281 @@
+	return u.UnmarshalLogs(buf)
+}
+
+func (u *CloudTrailLogsUnmarshaler) UnmarshalLogs(buf []byte) (plog.Logs, error) {
```
By the way, the CloudTrail log files are also compressed (see [documentation](https://docs.aws.amazon.com/awscloudtrail/latest/userguide/cloudtrail-log-file-examples.html)). So this unmarshaler would need the same `sync.Pool` of gzip readers the others are using too.

I think this part could be removed from the unmarshalers in the future so we could just use that code one time, but for now we have to keep repeating it.

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/cloudtraillogs/cloudtraillogs.go:
```diff
@@ -0,0 +1,283 @@
+	}
+
+	var record CloudTrailRecord
+	if err := json.Unmarshal(decompressedBuf, &record); err != nil {
```
Does the `decompressedBuf` contain only one cloudtrail record? Just double-checking since other JSON compressed logs are usually split by new line (so each line would be a record)

jsvd replied:
switched the implementation to read the {"Records": [...]} object and process all the records, grouping records by resource (cloud id/cloud region)

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/cloudtraillogs/cloudtraillogs.go:
```diff
@@ -0,0 +1,283 @@
+	EventType                    string         `json:"eventType"`
+	EventCategory                string         `json:"eventCategory"`
+	RecipientAccountID           string         `json:"recipientAccountId"`
+	UserIdentity                 map[string]any `json:"userIdentity"`
```
I think we should try to remove any use of `any`. They are quite expensive in performance. I see, for example, for user identity we only care about `principalId`, `userName` and `arn`, and they are all strings, so maybe we could even replace this with

```
UserIdentity {
  PrincipalID string `mapstructure: principalId`
  ...
} `json:"userIdentity"`
```

What do you think?

And then the same would apply for any other `any`s.

jsvd committed:
introduce benchmark and switch to go-json

jsvd commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/cloudtraillogs/cloudtraillogs.go:
```diff
@@ -0,0 +1,283 @@
+	}
+
+	var record CloudTrailRecord
+	if err := json.Unmarshal(decompressedBuf, &record); err != nil {
```
done

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/cloudtraillogs/cloudtraillogs.go:
```diff
@@ -0,0 +1,308 @@
+			return plog.Logs{}, fmt.Errorf("failed to decompress CloudTrail logs: %w", errDecompress)
+		}
+	} else {
+		// Content is not gzip-compressed, use the original buffer
```
Can this happen for cloud trail logs? I thought cloud trail logs sent to S3 had to be compressed, unless it was some kind of manual upload, but maybe I am wrong

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/cloudtraillogs/cloudtraillogs.go:
```diff
@@ -0,0 +1,308 @@
+	logs := plog.NewLogs()
+
+	// Group records by account ID and region to create separate resource logs
+	groupedRecords := make(map[string][]CloudTrailRecord)
+	for _, record := range records {
+		// Use account ID and region as the key for grouping
+		key := record.RecipientAccountID + "|" + record.AwsRegion
+		groupedRecords[key] = append(groupedRecords[key], record)
+	}
```
From the [documentation](https://docs.aws.amazon.com/awscloudtrail/latest/userguide/cloudtrail-log-file-examples.html) it says:

> CloudTrail uses the following file name format for the log file objects that it delivers to your Amazon S3 bucket:
> AccountID_CloudTrail_RegionName_YYYYMMDDTHHmmZ_UniqueString.FileNameFormat 

In that case, would it be correct to assume all logs coming from the same S3 file (I assume this is the case) have the same account ID and region?

jsvd committed:
simplify code to assume compressed and same id/region

also refactor test code to reuse what the collector provides

jsvd commented:
The code has been simplified to assume compressed data and same account id/region. test code was also dramatically reduced by reusing what the collector provides to compare logs, ignoring order of records.

jsvd committed:
fix import according to CI

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/cloudtraillogs/cloudtraillogs.go:
```diff
@@ -0,0 +1,290 @@
+	}
+
+	var cloudTrailLogs CloudTrailLogs
+	if err := json.Unmarshal(decompressedBuf, &cloudTrailLogs); err != nil || len(cloudTrailLogs.Records) == 0 {
```
Maybe use `gojson`?

Why would `len(cloudTrailLogs.Records) == 0` be an unmarshaling error?

jsvd replied:
changed to gojson and agreed that the check doesn't belong here.

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/cloudtraillogs/cloudtraillogs.go:
```diff
@@ -0,0 +1,290 @@
+func (u *CloudTrailLogsUnmarshaler) processRecords(records []CloudTrailRecord) plog.Logs {
+	logs := plog.NewLogs()
+
+	if len(records) == 0 {
+		return logs
+	}
```
I think that it makes sense to have the validation here instead of the unmarshaling one. WDYT? Relates to my previous comment

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/cloudtraillogs/cloudtraillogs.go:
```diff
@@ -0,0 +1,290 @@
+}
+
+// checks if the event is related to EC2 instance operations
+func isEC2InstanceOperation(eventName string) bool {
+	ec2Operations := map[string]bool{
+		"StartInstances":     true,
+		"StopInstances":      true,
+		"TerminateInstances": true,
+		"RebootInstances":    true,
+		"RunInstances":       true,
+	}
+	return ec2Operations[eventName]
+}
```
Not to be a pain, but maybe just use `switch` instead of a map lookup? :)  It is better for performance, but I doubt it would should on an overall benchmark

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/cloudtraillogs/cloudtraillogs.go:
```diff
@@ -0,0 +1,290 @@
+}
+
+func (u *CloudTrailLogsUnmarshaler) setLogRecord(logRecord plog.LogRecord, record CloudTrailRecord) {
+	if t, err := time.Parse(time.RFC3339, record.EventTime); err == nil {
```
What if there is an error? Should it fail? Do you know what happens if we don't set a timestamp on a log record?

jsvd replied:
Given this is part of the cloudtrail platform itself and not timestamps coming from user's code it is very unlikely this parsing would fail, but I changed the behavior to resemble s3-access-log for consistency and added a test.

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/cloudtraillogs/cloudtraillogs.go:
```diff
@@ -0,0 +1,290 @@
+func extractInstanceDetails(responseElements map[string]any) []map[string]string {
+	var instanceDetails []map[string]string
+
+	if instancesSet, ok := responseElements["instancesSet"].(map[string]any); ok {
+		if items, ok := instancesSet["items"].([]any); ok {
```
I think that doing this for every record and using `any` is not ideal, but if you feel strongly that using it like this as a starter is not an issue, we can check this again on a follow up

jsvd replied:
I'd rather handle these anys at a later stage.

constanca-m replied:
I am fine doing this as a follow-up. But the differences of switching are quite clear on benchmarks.

For example, I switched your `UserIdentity` struct to this:
```
	UserIdentity struct {
		PrincipalId string `json:"principalId"`
		Arn         string `json:"arn"`
		UserName    string `json:"userName"`
	} `json:"userIdentity"`
```

Then fixed the way it was on use on attributes:
```diff
-       if record.UserIdentity != nil {
-               if principalID, ok := record.UserIdentity["principalId"].(string); ok {
-                       attrs.PutStr("principal.id", principalID)
-               }
-               if userName, ok := record.UserIdentity["userName"].(string); ok {
-                       attrs.PutStr("principal.name", userName)
-               }
-               if arn, ok := record.UserIdentity["arn"].(string); ok {
-                       attrs.PutStr("principal.iam.arn", arn)
-               }
+       if record.UserIdentity.PrincipalId != "" {
+               attrs.PutStr("principal.id", record.UserIdentity.PrincipalId)
+       }
+       if record.UserIdentity.UserName != "" {
+               attrs.PutStr("principal.name", record.UserIdentity.UserName)
+       }
+       if record.UserIdentity.Arn != "" {
+               attrs.PutStr("principal.iam.arn", record.UserIdentity.Arn)
        }
```

And compared the benchmarks:
```
goos: linux
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/extension/encoding/awslogsencodingextension/internal/unmarshaler/cloudtraillogs
cpu: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz
                           │   old.txt    │             current.txt              │
                           │    sec/op    │    sec/op     vs base                │
UnmarshalLogs/1_log-16       38.25µ ± 18%   34.94µ ± 16%        ~ (p=0.105 n=10)
UnmarshalLogs/10_logs-16     289.4µ ±  7%   250.2µ ±  6%  -13.57% (p=0.000 n=10)
UnmarshalLogs/100_logs-16    2.794m ± 10%   2.347m ±  8%  -16.00% (p=0.000 n=10)
UnmarshalLogs/1000_logs-16   29.44m ±  7%   20.99m ±  6%  -28.71% (p=0.000 n=10)
geomean                      976.9µ         810.0µ        -17.08%

                           │    old.txt    │              current.txt              │
                           │      B/s      │      B/s       vs base                │
UnmarshalLogs/1_log-16       20.97Mi ± 15%   20.00Mi ± 14%        ~ (p=0.739 n=10)
UnmarshalLogs/10_logs-16     3.281Mi ±  8%   3.304Mi ±  6%        ~ (p=0.643 n=10)
UnmarshalLogs/100_logs-16    683.6Ki ± 11%   722.7Ki ±  8%        ~ (p=0.185 n=10)
UnmarshalLogs/1000_logs-16   351.6Ki ±  8%   439.5Ki ±  7%  +25.00% (p=0.000 n=10)
geomean                      1.993Mi         2.115Mi         +6.15%

                           │   old.txt    │             current.txt              │
                           │     B/op     │     B/op      vs base                │
UnmarshalLogs/1_log-16       18.77Ki ± 0%   14.86Ki ± 0%  -20.86% (p=0.000 n=10)
UnmarshalLogs/10_logs-16     195.1Ki ± 0%   177.9Ki ± 0%   -8.80% (p=0.000 n=10)
UnmarshalLogs/100_logs-16    1.972Mi ± 0%   1.803Mi ± 0%   -8.56% (p=0.000 n=10)
UnmarshalLogs/1000_logs-16   21.18Mi ± 0%   17.91Mi ± 0%  -15.46% (p=0.000 n=10)
geomean                      632.8Ki        546.9Ki       -13.57%

                           │   old.txt   │             current.txt             │
                           │  allocs/op  │  allocs/op   vs base                │
UnmarshalLogs/1_log-16        217.0 ± 0%    176.0 ± 0%  -18.89% (p=0.000 n=10)
UnmarshalLogs/10_logs-16     1.994k ± 0%   1.594k ± 0%  -20.06% (p=0.000 n=10)
UnmarshalLogs/100_logs-16    19.66k ± 0%   15.66k ± 0%  -20.36% (p=0.000 n=10)
UnmarshalLogs/1000_logs-16   196.1k ± 0%   156.1k ± 0%  -20.40% (p=0.000 n=10)
geomean                      6.391k        5.117k       -19.93%
```

constanca-m reviewed (commented):
Thanks @jsvd ! Looks good. I am not commenting on the attribute names, I rather leave that part for @axw.

A couple of more comments:

jsvd committed:
changes related to review feedback

* ensure go-json is used everywhere
* handled failed timestamp parsing and added test
* refactor logic for empty records and added test

constanca-m commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/cloudtraillogs/testdata/cloudtrail_logs_expected.yaml:
```diff
@@ -0,0 +1,125 @@
+              - key: principal.iam.arn
+                value: { stringValue: arn:aws:iam::123456789012:user/Mateo }
+              - key: tls.protocol.version
+                value: { stringValue: TLSv1.2 }
```
Should the version just be `1.2`? We do the same for the remaining unmarshalers

jsvd committed:
strip TLSv from tls.protocol.version

constanca-m reviewed (commented):
It looks ready to me. I will let @axw give the approval, since I am in doubt in the attribute names

constanca-m reviewed (commented):
@jsvd I remembered something that is missing:
- Add a new config in `testdata/config.yaml` with the new format and update `config_test.go`
- Add new test in `extension_test.go` for this format

jsvd committed:
add config/config_test for cloudtrail

axw reviewed (commented):
@jsvd sorry for leaving this for so long. Predictably, most of my comments are about SemConv.

Aside from that, I'm uneasy about the method-specific translation. Although it would be a nicer result for those methods, IIANM we won't capture the request/response for any other methods - that doesn't seem ideal.

If nothing else, I feel like we should capture request/response for everything consistently, and then we could perhaps enrich for certain services/methods. WDYT?

jsvd committed:
changes to reflect review

* remove response/request specific processing
* rename request.event_id to aws.cloudtrail.event_id
* rename principal.iam.arn to principal.arn

jsvd committed:
fix README

jsvd committed:
fix oopses

jsvd committed:
rename files

jsvd commented on extension/encoding/awslogsencodingextension/README.md:
```diff
@@ -128,3 +136,44 @@ extensions:
+| CloudTrail field                                    | Attribute in OpenTelemetry log                                                                             |
+|-----------------------------------------------------|------------------------------------------------------------------------------------------------------------|
+| `requestParameters.instancesSet.items[].instanceId` | `aws.request.parameters.instances` (array of instance IDs)                                                 |
+| `responseElements.instancesSet.items[]`             | `aws.response.instances` (array of instance details including instanceId, currentState, and previousState) |
```
done

jsvd commented on extension/encoding/awslogsencodingextension/README.md:
```diff
@@ -128,3 +136,44 @@ extensions:
+|--------------------------------|-----------------------------------------------------------------------------------------------|
+| `userIdentity.principalId`     | `principal.id` (if available)                                                                 |
+| `userIdentity.userName`        | `principal.name` (if available)                                                               |
+| `userIdentity.arn`             | `principal.iam.arn` (if available)                                                            |
```
moved to principal.arn

jsvd commented on extension/encoding/awslogsencodingextension/README.md:
```diff
@@ -128,3 +136,44 @@ extensions:
+|--------------------------------|-----------------------------------------------------------------------------------------------|
+| `userIdentity.principalId`     | `principal.id` (if available)                                                                 |
+| `userIdentity.userName`        | `principal.name` (if available)                                                               |
+| `userIdentity.arn`             | `principal.iam.arn` (if available)                                                            |
```
removed their separation from all other events

jsvd commented:
I believe I have addressed most comments:
* a few field renames
* removed special handling of certain response elements and request parameters
I've also rename the files to adhere to the rest of the unmarshalers.

jsvd committed:
added more tests

jsvd committed:
fix benchmark

jsvd commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/cloudtraillogs/unmarshaler.go:
```diff
@@ -0,0 +1,273 @@
+}
+
+func (u *CloudTrailLogsUnmarshaler) UnmarshalLogs(buf []byte) (plog.Logs, error) {
+	gzipReader, ok := u.gzipPool.Get().(*gzip.Reader)
```
done

jsvd committed:
support cloudtrail on extension

axw reviewed (commented):
Thanks @jsvd. Just a couple more things, then it's good to go. I'd like to come back to some of the other attribute names, but we can do it in a followup.

jsvd committed:
add a few more fields plus schema tweaks

jsvd committed:
update readme

jsvd commented on extension/encoding/awslogsencodingextension/README.md:
```diff
@@ -167,4 +173,33 @@ extensions:
+| `eventSource`                         | `rpc.service`                                                                          |
+| `eventName`                           | `rpc.method`                                                                           |
+| `requestID`                           | `aws.request_id`                                                                       |
+| `sourceIPAddress`                     | `net.peer.ip`                                                                          |
```
done

axw reviewed (approved):
Thank you @jsvd!

axw commented on extension/encoding/awslogsencodingextension/README.md:
```diff
@@ -167,4 +173,33 @@ extensions:
\ No newline at end of file
+| `oversizeFields`              | _Currently not supported_                                                                        |
+
+#### CloudTrail log record fields
```
Thank you, I agree with taking that as follow-ups

edmocosta commented on extension/encoding/awslogsencodingextension/internal/unmarshaler/cloudtraillogs/unmarshaler.go:
```diff
@@ -0,0 +1,230 @@
+	}
+
+	if record.UserIdentity != nil {
+		if principalID, ok := record.UserIdentity["principalId"].(string); ok {
```
Any security concern about adding those identity fields to the logs attributes by default? It seems cloudtrail redacts the `userName` already when it's necessary, but I'm not sure about the other ones.

edmocosta replied:
https://aws.amazon.com/blogs/security/modifications-to-aws-cloudtrail-event-data-of-iam-identity-center/#:~:text=Effective%20July%2014%2C%202025%2C%20IAM,provides%20guidance%20for%20these%20updates.

axw replied:
My 2c: I don't think it's the role of the encoding extension to redact things. If a user wants to redact, they can use the filter or transform processors for example.

edmocosta replied:
Yes I agree, I didn't mean redacting the values, just avoiding adding them to the log attributes. Anyway, If you folks think they are not a security concern, I guess we can keep it as it's and just add the extra IAM fields.

jsvd committed:
drop principalId and userName in favor of userId and identityStoreArn

jsvd committed:
rename cloudtrail_logs to cloudtrail_log

jsvd commented on extension/encoding/awslogsencodingextension/config.go:
```diff
@@ -17,6 +17,7 @@ const (
 	formatVPCFlowLog                       = "vpc_flow_log"
 	formatS3AccessLog                      = "s3_access_log"
 	formatWAFLog                           = "waf_log"
+	formatCloudTrailLogs                   = "cloudtrail_logs"
```
done

jsvd committed:
reinstate principalId and userName

jsvd committed:
renaming

jsvd committed:
leftover renames for cloudtraillog instead of cloudtraillogs

jsvd committed:
update readme table

jsvd commented:
A few more updates:
* cleaned up last renaming references of "cloudtraillogs" in favor of "cloudtraillog", including file renames
* updated readme with recent schema changes
* reinstated principalId and userName, populate with whatever the event brings leaving redaction to the processing stage.
* rebased against master due to the gzip fix

edmocosta reviewed (approved):
LGTM, @axw @constanca-m could you please have a final look at the latest changes so we can mark it as ready to merge. Thank you!

axw reviewed (approved):
Looks even better :) Thanks!

edmocosta committed:
Merge branch 'main' into add_cloud_trail_decoding

jsvd committed:
Merge branch 'main' into add_cloud_trail_decoding
